### ISSUE TITLE
Slow tab-completion on large data object in notebook

### ISSUE THREAD

**Issue by mangecoeur** [ipython/ipython#10493]
When using ipython6 with Jupyter notebook, I'm noticing much longer delays for tab-completion on large variables (arrays ~1GB in size in memory) which don't occur with ipython5.3 .

**Comment by takluyver** [comment]
Does `repr(x)` with these arrays take a long time? A similar issue has been reported as jupyter/qtconsole#90. The completion code winds up indirectly calling repr on the objects. We're not sure how to avoid that at present.

**Comment by mangecoeur** [comment]
`repr` is somewhat slow on these, but on some datasets the tab completion was much slower than `repr` and in at least one case froze the notebook (and if i remember rightly also used a ton of memory- I don't have time to re-upgrade and check). This was caused when using geopandas to load a global shapefile of countries (http://www.naturalearthdata.com/downloads/10m-cultural-vectors/10m-admin-0-countries/) and triggering tab-completion - note that now I look at it again, it's not actually a big file.

I imagine something is trying to recursively scan the object, possibly getting caught up in a circular reference somewhere.

**Comment by Carreau** [comment]
Jedi may be inspecting things even if its results don't show up in the notebook. There is a timeout that should stop jedi to inspect things after ~0.8 seconds 9as soon as jedi give us a change to cancel it), and jedi have its on configuration options to change its agressiveness at inspecting.  IPython should behave if jedi is not installed so you can uninstall it  to see if it's the cause.

Jedi master should also be way  faster than stable, and keep in mind that once Jedi has had a change to completely inspect a module it will cache the result, so completion should be faster, after a couple of tab press.

**Comment by makmanalp** [comment]
I'm having this exact issue with large pandas dataframes. As described, repr does take some time but not nearly as much time as how much it's lagging, perhaps repr gets re-called on each keypress?

This seems to be happening in conjunction with https://github.com/ipython/ipython/issues/9879 but I'm unclear on whether they're related (see more in that ticket).

This is with `jedi==0.10.2`. When I remove jedi, everything is as smooth as butter, suggesting the issue is there. When I install jedi from HEAD with `pip install -e git://github.com/davidhalter/jedi.git#egg=jedi` (currently at commit e2f88db3c227965179d472f5e0d6a07b34e6f364) it's still quite slow.

Linking to related issue in jedi: https://github.com/davidhalter/jedi/issues/919

**Comment by makmanalp** [comment]
In the meantime, is there a way I can disable jedi in ipython with `Completer.use_jedi`: https://ipython.readthedocs.io/en/stable/config/options/terminal.html

**Comment by michaelaye** [comment]
I'm having the same issue of very very slow object inspection via TAB, and after applying fix, it's no issue at all (using current conda-forge versions on a current macOS).
My object that I inspect is created by this code, in case somebody wants to check something:

https://github.com/michaelaye/planet4/blob/master/planet4/io.py#L452-L572

**Comment by takluyver** [comment]
@michaelaye  I think that may be a separate issue, because it looks like the `repr()` of your object should be quick to calculate. Maybe its invoking properties? It looks like some of those properties query the file.

If you've got the time to spare, please do talk to @Carreau, who's taking the lead on the jedi integration, so that we can work out a proper fix for that issue - we're adding jedi integration with the eventual aim of replacing our own completion-finding machinery, so turning jedi off is strictly a short-term solution. It *shouldn't* be calling properties like that.

**Comment by Carreau** [comment]
> If you've got the time to spare, please do talk to @Carreau, who's taking the lead on the jedi integration, so that we can work out a proper fix for that issue - we're adding jedi integration with the eventual aim of replacing our own completion-finding machinery, so turning jedi off is strictly a short-term solution. It shouldn't be calling properties like that.

I'm more and more finding case where jedi is not sufficient. I think I'm going to postpone the removal of our machinery. I need to find time to dive into jedi, but it is also a complex beast. I've already spoted a couple of issues I need to fix the correct way. In particular indeed, it does every now and then trigger dynamic properties.

**Comment by takluyver** [comment]
:-( That's a pity. I'll leave it up to you whether the problems are serious enough that we should disable it by default and let experimental users play with it.

**Comment by Carreau** [comment]
> That's a pity. I'll leave it up to you whether the problems are serious enough that we should disable it by default and let experimental users play with it.

it's a bit of a catch 22, if we disable it by default, we don't get bug reports.

**Comment by pwaller** [comment]
Possibly related issue: https://github.com/davidhalter/jedi/issues/931

**Comment by MrWaterZhou** [comment]
seems it has something to do with pandas. I tried a large numpy.array and auto-completion is smooth; but 10s for a pandas.DataFrame with the same data.

**Comment by joonro** [comment]
I was having this issue as well and it seems updating `jedi` to 0.11.0 from 0.10.2 solved this for me. It would be great if others can confirm this. Thank you!

**Comment by dkapitan** [comment]
@joonro 
Yes, updating to `jedi==0.11` worked for me as well!

**Comment by michaelaye** [comment]
Incidentally, is it actually impossible to get tab completion while the notebook is waiting for a cell to execute?

**Comment by takluyver** [comment]
Yes-ish. The completion requests go into the same queue as execution requests, so they won't get a response while code is executing. The frontend has some very simple completion machinery to pick up words that you've already used, so you should still see those even if the kernel is busy.

**Comment by makmanalp** [comment]
`jedi==0.11` definitely makes it way faster, it's now OK for most normal cases except when I do `df.<tab>` or something and the resulting list is large, then it takes some time. But from my perspective it's now eminently usable 🎉

**Comment by rusiano** [comment]
I am having a very simiIar/related issue: the shift-tab documentation inspector is extremely slow. I tried to follow your steps (like the jedi thing == False, I also updated jedi to 0.11) hoping to find a solution but I am still facing this annoying problem.

The things I noticed are:
- This morning I was not having this issues: I started noticing it after I shut down the notebook when I was having a cell running forever (doing like a join between two pandas dataframe - sized roughly 17MB 
each). 
- The shift-tab works fine if I call it for functions called from a small object like a list (e.g. lst.apply() is ok), but as soon as I call it from my "big" dataframe it gets stuck (e.g. df['Column1'].apply() is a problem).
- The documentation inspector shows immediately if I manually interrupt the Kernel, suggesting that the documentation was ready but the kernel was still running for unknown reasons.

I hope someone can help me.

This is some information about my environment.

> {'commit_hash': 'ca5443062',
 'commit_source': 'installation',
 'default_encoding': 'cp1252',
 'ipython_path': 'C:\\Users\\rusia\\Anaconda3\\lib\\site-packages\\IPython',
 'ipython_version': '6.2.1',
 'os_name': 'nt',
 'platform': 'Windows-10-10.0.16299-SP0',
 'sys_executable': 'C:\\Users\\rusia\\Anaconda3\\python.exe',
 'sys_platform': 'win32',
 'sys_version': '3.6.1 |Anaconda 4.4.0 (64-bit)| (default, May 11 2017, '
                '13:25:24) [MSC v.1900 64 bit (AMD64)]'}

**Comment by ElieGouzien** [comment]
Could you try with python 3.7 ?

It's likely that https://github.com/python/cpython/pull/2132 helps.

You also may have a look at https://github.com/davidhalter/jedi/pull/922

**Comment by chorus12** [comment]
In jupyter just run a cell with the following contents `%config IPCompleter.use_jedi = False`

**Comment by LunarLanding** [comment]
I am at jedi 0.13.3 and this is still a serious issue that considerably slows down my coding. I often have to learn new stuff. I disabled jedi for now, but the hints it gave are more complete that the default autocomplete.

**Comment by michaelaye** [comment]
Same here. More than once jedi slowed down my system and I needed to hunt and/or remember that this could be it. So I just learned to live without it for now.

**Comment by yogevyuval** [comment]
What's the status on this one? Is Jedi just not recommenced for big datasets? is this planned to be fixed some day?

**Comment by Carreau** [comment]
> is this planned to be fixed some day?

Currently I am – in practice –  the sole maintainer of IPython, and this is not my full time job. 
I have no current plan to personally fix this myself. I do not have the time to look into this, nor the funds to hire someone to do it. So well...wether it will be fixed is up to the community.

Currently my priorities are :
1) Be nice and try to at least respond to every new issues and PR. 
   - I'll prioritize being nice(er) to new comers, and minorities, as well as simple and strait-forward questions and PRs. I rarely have 2h block I can devote  on complex things. 
2) Maintainability, documentation and regular release of the project.
    - Mostly this ends up being 1 friday afternoon per month to do release notes and actual release.

The rest is well... up for grabs, and can depends on what I want to work on.

We are of course always looking for more involved contributors to help maintain the project, and looking for a path to make them core dev with commit rights.

**Comment by ElieGouzien** [comment]
@michaelaye @yogevyuval 

I don't have time to look at that bug, but i did fixed something close without any specific knowledge on IPython or jedi.
My advises would be:
1. Make sure you have recent versions of IPython, Jedi and Python (3.7).
2. Find a minimal example.
   If it implies `__repr__()` you can use this kind of class to identify it:
    ```python

        class SlowRepr():
            "class to test what happens if __repr__ is very slow."
            def some_method(self):
                pass
            def __repr__(self):
                time.sleep(2)
    ```
3. Once you are sure of the exact cause, replace the `time.sleep()` by `pdb.set_trace()` or `traceback.print_stack()` to locate where `__repr__()` is called.
4. Understand why you ended here (this is the hard part), and figure out if it's an unwanted behavior.
5. Make a fix, and propose a pull request.
